### PR TITLE
Stomach given its own reagent_container

### PR DESCRIPTION
**Pull request by flagzzzz**
Gives the stomach its own reagent_container that slowly "digests" reagents within it and transfers them to the body. Having the stomach interact with the body's whole reagent_container seemed odd, especially since removing the stomach effectively limited the body's whole reagent capacity to 25 or so unit.

This digestion rate is generally 1u per `process()` call (about once a second) split across all of the different reagents in the stomach. Individual reagents can have different digestion rates and more room is given for upgraded stomachs that could be given different reagent capacity, digestion rate, and possibly filters for reducing the amount of toxic reagents that make their way to the body, etc..

Also adds a proc for reagents called `digest()` that lets reagents do things in the stomach. Charcoal has been updated to use this. For the purposes of overdose calculations, the volume of a reagent in the body can be added to the volume of the same reagent in the stomach if a var for the reagent if `overdose_includes_stomach = TRUE`. ~~None~~ All of the current reagents have this var set to `TRUE` but it may come in handy in the future.

Any feedback or ideas are welcome.

Reagents that cause damage to or heal the stomach:

| Reagent | Damage (per second) | Stomach damage threshold (u) |
| --- | --- | --- |
| ~~Anti-Toxin~~ | ~~0.1~~| ~~15~~ |
| Bleach/spacecleaner/fishbleach | 1 | >0 |
| Ethanol | 0.08 | >0 |
| Imidazoline | -0.5 (heals) | >0 |

Other potential ideas/expansions:
* Give the stomach a number of bites variable that tracks how much food is in it. This value, and maybe even stomach damage, feeds into fullness calculations.
* Robotic/advanced stomach
* food/reagent temp causing damage
* stomach damage causing blood to be transferred to the stomach which leads to bloody puke

:cl:
* rscadd: The stomach holds reagents separately from the body and slowly transfers its contents to the body's reagent container. Individual reagents can interact with the stomach. THIS ONLY HAPPENS WITH DRINKING CURRENTLY.
* tweak: Some chemicals now damage or heal the stomach. Imidazoline heals it. Ethanol, Bleach, and Antitoxin damage it.

### CODE REVIEW

**Comment by MadmanMartian** [comment]
Just what I wanted to do but was too scared to.
👏

**Comment by ShiftyRail** [comment]
>Antitoxin damages stomach

Ah yes, perfectly what I needed for my kill and clone routine :clap:

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 1]
```diff
@@ -579,6 +603,9 @@
 	density = 1.49033
 	specheatcap = 0.55536
 
+/datum/reagent/anti_toxin/digest(var/mob/living/carbon/human/M)
+	..(M, damage = 0.1, amount_for_damage = 15)
```
Comment (ShiftyRail) [inline code/modules/reagents/Chemistry-Reagents.dm:607]: I feel it would be more elegant to have these be vars on the reagent.

**Comment by Dacendeth** [comment]
How does this affect races without a stomach?

**Comment by flagzzzz** [comment]
They aren't affected by this at all.

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
So what happens with reagents like chloral? Now it takes 10 seconds to reach 10u in blood, during which 2u were already consumed.

And also takes 30 or more seconds to overdose anyone on anything

**Comment by flagzzzz** [comment]
>So what happens with reagents like chloral? Now it takes 10 seconds to reach 10u in blood, during which 2u were already consumed.

For that, the `on_mob_life` check could be modified to take the stomach volume of chloral hydrate into account. Either that or the digestion rate of chloral hydrate increased like I did for nanobots. Good point though.

>And also takes 30 or more seconds to overdose anyone on anything

There's a flag that can be turned on that includes the reagent volume in the stomach in overdose calculations. Currently, it's defaulted to off for all reagents and can be turned on by a per-reagent basis but if it's too much of a concern then it can be set to be on by default.

**Comment by ShiftyRail** [comment]
So not only does VERY HEAVILY encourages kill and clone (by making the processing of healing chems slower, and by hurting the tummy if you feed more than 15u anti-tox), this only affects humans ?

Terrible change tbh

File: code/modules/reagents/reagent_containers.dm [hunk 2]
```diff
@@ -349,8 +349,13 @@ var/list/LOGGED_SPLASH_REAGENTS = list(FUEL, THERMITE)
 			reagents.reaction(user, INGEST)
 			spawn(5)
 				if(reagents)
-					reagents.trans_to(user, amount_per_imbibe)
-
+					if(user.get_stomach())
+						var/datum/organ/internal/stomach/S = user.get_stomach()
+						reagents.trans_to(S.get_reagents(), amount_per_imbibe) // transfer to stomach
+					else if(istype(user, /mob/living/carbon/human))
+						reagents.trans_to(user, amount_per_imbibe / 4) // greatly reduced reagent absorption without a stomach
```
Comment (ShiftyRail) [inline code/modules/reagents/reagent_containers.dm:356]: You're dividing by 4 the effectiveness of all medicine, and poisons, to races without a stomach

Really ?

File: code/modules/reagents/reagents/vg.dm [hunk 3]
```diff
@@ -13,16 +13,15 @@ var/global/list/charcoal_doesnt_remove=list(
...
-/datum/reagent/charcoal/on_mob_life(var/mob/living/M)
-	if(!M)
-		M = holder.my_atom
-
-	if(ishuman(M) && prob(5))
-		var/mob/living/carbon/human/H=M
-		H.vomit()
+/datum/reagent/charcoal/digest(var/mob/living/carbon/human/M)
+	var/datum/organ/internal/stomach/S = M.get_stomach()
+	if(!S)
```
Comment (ShiftyRail) [inline code/modules/reagents/reagents/vg.dm:20]: You made charcoal not even work with persons/races without a stomach

File: code/modules/reagents/reagent_containers.dm (hunk 2 above)
Comment (flagzzzz) [reply code/modules/reagents/reagent_containers.dm:356]: The intention was to limit the effectiveness of medicine if you were a member of a race that had a stomach but didn't have a stomach as an individual. Thanks for pointing that out, I'll correct it.

Comment (ShiftyRail) [reply code/modules/reagents/reagent_containers.dm:356]: All good

**Comment by Dacendeth** [comment]
it doesnt only affect humans, greys and vox have stomachs too.

**Comment by boy2mantwicethefam** [comment]
bruh how tf is kill and healing real just inject them instead of making them swallow pills

**Comment by strawberrytan** [comment]
I don't know how to feel about these changes but I can suggest the chemicals found in dan foods to do low stomach damage for the theme

**Comment by kilozombie** [comment]
I will say it doesn't make that much sense for ethanol to damage your stomach

**Comment by MadmanMartian** [comment]
This stomach seems to only be used from drinking from a container of some form. Is this intended?

**Comment by Irradiation** [comment]
I'm going to go on a limb and assume the OP intended this PR to work for _every_ method of chemical ingestion/intake/absorption/whatever, so here's my deux sous at figuring out this PR and its effects on medbay and chemistry:

**PROS**
-More pseudo realism by trying to simulate a stomach organ.
-Takes longer to get drunk, due to Ethanol having a longer dwell time before hitting your system.

**NEUTRAL** (Depends on your point of view)
-You can't ingest large amounts of chemicals in general or your tummy is going to damage you.
-Side effect of chemical healing speed is that it also nerfs (speed)cloning by slowing down the intake of chemicals from cryo.

**CONS**
-Chemical healing speed has been slowed down through taking 2 seconds for 1 unit to start taking effect. **Depending on how bad this slows down chemical healing, people might unironically kill-and-clone because they might as well make a new body. Or head-swap with a body that's laying around.**
-Alcoholic drinks (I presume they count as Ethanol for the purpose of digestion) will double-whammy the hell out of your liver AND stomach.
-Potentially neuters intentional overdoses (Beneficial or not). **That's going to take a ton of fiddling due to the amount of chemicals that have an overdose and will very probably cause oversights, but that might be salvagable.**
-Chloral Hydrate, along with any other chemical that requires a certain amount of units before they cause certain effects, are going to take a shotgun blast to the proverbial testicles. **If you're going through with this, give those chemicals a faster disgestion rate**.

**Comment by Reviire** [comment]
So is there a way to bypass this, or? Because if injections go into the stomach, that's pretty retarded.

Now, while I don't mind this PR, the issue is, everything we have that's chemical related isn't designed for this. You've got a lot more to do if you want to make any of that stuff work like normal.

**Comment by Reviire** [comment]
Gonna throw this idea out there. Make reagent digestion ramp up over time. So large amounts of reagents don't take literally years to go through. let it start off slow, and the reagent should digest faster based on how long it's been in. So large amounts of reagents don't just cease existing. 

Also make literally everything function at least somewhat like it used to, don't kill the chloral/overdose/every medicine/every poison.

**Comment by MadmanMartian** [comment]
@Reviire A quick look at the code shows that the stomach's only used by drinking from a reagent container.

**Comment by MusgardTiger** [comment]
please get rid of the ethanol meme @flagzzzz

**Comment by SonixApache** [comment]
somewhat okay idea, terrible execution

**Comment by bathosbathtime** [comment]
What are these strait edge memes with ethanol based drinks nuking you so hard.  Why must we make the bartender so lonely.

**Comment by nuklearcellphoneg** [comment]
Life just keeps getting better and better.

**Comment by kilozombie** [comment]
@jknpj Apparently this PR only affects drinks instead of pills which seems unintended

Also the fact that ethanol still damages your stomach is really stupid, not sure this should have been merged yet/ever

**Pull request merged**